File: nearest_exit_maze.rs
use std::collections::VecDeque;
use std::collections::HashSet;
struct Solution;

impl Solution {
    pub fn nearest_exit(maze: Vec<Vec<char>>, entrance: Vec<i32>) -> i32 {
        let (mut dq, mut visited) = (VecDeque::new(), HashSet::new());
        dq.push_back((entrance[0] as usize, entrance[1] as usize, 0));
        while !dq.is_empty() {
            while let Some((i, j, c)) = dq.pop_front() {
                if !visited.insert((i, j)) {
                    continue;
                }
                if i == 0 || i == maze.len() - 1 || j == 0 || j == maze[0].len() - 1 {
                    if i != entrance[0] as usize || j != entrance[1] as usize {
                        return c;
                    }
                }
                if maze[i][j] == '.' {
                    if i > 0 && maze[i - 1][j] == '.' {
                        dq.push_back((i - 1, j, c + 1));
                    }
                    if i < maze.len() - 1 && maze[i + 1][j] == '.' {
                        dq.push_back((i + 1, j, c + 1));
                    }
                    if j > 0 && maze[i][j - 1] == '.' {
                        dq.push_back((i, j - 1, c + 1));
                    }
                    if j < maze[0].len() - 1 && maze[i][j + 1] == '.' {
                        dq.push_back((i, j + 1, c + 1));
                    }
                }
            }
        }
        -1
    }
}

File: other/nearest_exit_maze.cpp
#include "leetcode.h"

class Solution {
public:
  int nearestExit(vvd(char) & maze, vector<int> &entrance) {
    int m = maze.size(), n = maze[0].size();
    int dirs[5] = {0, -1, 0, 1, 0};
    auto isExit = [&](int i, int j) -> int {
      return (i != entrance[0] || j != entrance[1]) &&
             (i * j == 0 || i == m - 1 || j == n - 1);
    };
    deque<array<int, 3>> dqa = {{entrance[0], entrance[1], 0}};
    while (!dqa.empty()) {
      auto [i, j, s] = dqa.front();
      dqa.pop_front();
      for (int d = 0; d < 4; ++d) {
        int i2 = i + dirs[d];
        int j2 = j + dirs[d + 1];
        if (i2 < 0 || i2 >= m || j2 < 0 || j2 >= n || maze[i2][j2] == '+')
          continue;
        maze[i2][j2] = '+';
        if (isExit(i2, j2))
          return s + 1;
        dqa.push_back({i2, j2, s + 1});
      }
    }
    return -1;
  }
};

int main() {
  Solution obj;
  vvd(char)
      maze = {{'+', '+', '.', '+'}, {'.', '.', '.', '+'}, {'+', '+', '+', '.'}};
  vector<int> entrance = {1, 2};
  cout << obj.nearestExit(maze, entrance);
}
